### ISSUE TITLE
Feature request: batch images

### ISSUE THREAD

**Issue by agawrylak** [blib-la/runpod-worker-comfy#28]
Is it possible to add an option to return a batch of images?

**Comment by aleph65** [comment]
Was also looking for this

I guess for now workaround is to do the batch image logic locally... change seed as required

**Comment by aleph65** [comment]
yes exactly!

**Comment by TimPietrusky** [comment]
@aleph65 would you mind providing me with an example workflow, so that I use this for testing & development?

**Comment by vesper8** [comment]
I'm also interested in this. Just using the default test input from https://raw.githubusercontent.com/blib-la/runpod-worker-comfy/main/test_input.json

If I increase the batch_size from 1 to say 5, how can I get the 5 images separately from the API response?

**Comment by TimPietrusky** [comment]
> Just using the default test input from https://raw.githubusercontent.com/blib-la/runpod-worker-comfy/main/test_input.json
> 
> If I increase the batch_size from 1 to say 5, how can I get the 5 images separately from the API response?

Ok  thank you! I imagine we would deliver an array of images then. But we have to make sure that this is not breaking the current behavior OR release a breaking change.

**Comment by aleph65** [comment]
sorry for not sending this yet, my examples require additional packages and I want to send a out-of-box one

Basically, there are 2 scenarios:

1 - there is a batch of images in a Save Image or Display Image node

2 - There are multiple Save Image or Display Image nodes

If you query comfyui api today you will receive all of those outputs (all batches of all output image nodes)

Personally if backcompat is important you can just return the original first image in the same place and return the full array as well. And add a flag for people who have already migrated to this new signature

**Comment by RaresKeY** [comment]
I'm also looking for this feature.
Here's an example of the default flux workflow with changed batch latents to 2:
[workflow.json](https://github.com/user-attachments/files/16992381/workflow.json)

The way comfyui handles this is by saving 2 images sequentially, you guys send the last image in ["output"]["message"] so the easiest way is to read how many files where generated in folder before and after comfyui gen and send those as ["output"]["message1"], ["output"]["message2"] ... ["output"]["messageN"]

The reason I suggest reading the files instead of the workflow is in case the workflow contains multiple empty latents or custom nodes but for a default workflow you can just read "batch_size" to get the batch number.

this is some relevant part of a log from my runpod testing:
```
...
2024-09-13T10:07:20.027113249Z runpod-worker-comfy - image generation is done
2024-09-13T10:07:20.027156841Z runpod-worker-comfy - /comfyui/output/ComfyUI_00040_.png
...
2024-09-13T10:17:43.164717148Z runpod-worker-comfy - image generation is done
2024-09-13T10:17:43.164772808Z runpod-worker-comfy - /comfyui/output/ComfyUI_00042_.png
...
```

let me know if you need testing, I would be happy to help

**Comment by pribeh** [comment]
@RaresKeY your input seems sensible. also, thanks for sharing your work around. My suggestion here though would be to help provide the most optimal experience for the end-user who is already waiting for content to generate. Most platforms or projects have moved toward showing generated image as soon as they're ready to the end-user so that they can already get a sense of the results and reduce wait times. I would suggest uploading the images to S3 as soon as each are done rendering sequentially. 

I'm not certain if it would be best to get a response in each instance that an image is generated or just one response with the images in an array. That might still work okay as a single response. 

You can use a lambda function to trigger every time a file uploads to S3 to notify your server and/or client that the file is uploaded or you can poll the S3 asset until its ready as long as you know what the asset ID (filename) is.

**Comment by hlktrkn** [comment]
I apologize if I missed any updates about this request. The batch images output S3 upload would be an excellent enhancement. 
Thanks in advance @TimPietrusky

**Comment by TimPietruskyRunPod** [comment]
Batch images are supported through ComfyUI's standard \`batch_size\` parameter on \`EmptyLatentImage\`, \`EmptySD3LatentImage\`, etc. — set it to >1 and the workflow produces multiple images in one execution. The worker collects them all into the response \`images\` array.

For batching at the *request* level (multiple independent prompts in one request), that's not directly supported — submit them as separate \`/run\` calls and Runpod handles concurrency. The reason: each request gets a unique \`job_id\` for billing/tracking, mixing prompts in one request muddies that.

If you want to batch many independent prompts efficiently, set \`workersMax > 1\` and submit them in parallel — that gives you concurrency without the model-loading overhead of independent endpoints.

Closing as both forms are supported via existing patterns.

**Comment by RaresKeY** [comment]
Hey @TimPietruskyRunPod

I think the Issue was also for comfyui workflows containing multiple output images or desired outputs, like alpha PNG + PNG for example, was that also solved in the meantime?

Also another issue is that setting latent images to >1 increases VRAM usage (or did at the time I was using comfyui), so it limits our config surface.

Btw, I haven't worked with comfyui in a while or with this repo so I'm not up to date with changes since this Issue was opened, so these are more follow-up questions to consider rather than suggestions.
